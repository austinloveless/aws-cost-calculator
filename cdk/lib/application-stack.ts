import { Stack, Construct, CfnOutput, StackProps } from "@aws-cdk/core";
import { LambdaRestApi } from "@aws-cdk/aws-apigateway";
import {
  Function,
  Runtime,
  Code,
  IFunction,
  CfnParametersCode,
} from "@aws-cdk/aws-lambda";
import { ManagedPolicy, Role, ServicePrincipal } from "@aws-cdk/aws-iam";
import { Dashboard } from "@aws-cdk/aws-cloudwatch";
import { GraphWidget, Metric } from "@aws-cdk/aws-cloudwatch";

interface ApplicationStackProps extends StackProps {
  applicationName: string;
  stage: string;
}

export class ApplicationStack extends Stack {
  public readonly lambdaCode: CfnParametersCode;
  readonly handler: IFunction;

  constructor(scope: Construct, id: string, props: ApplicationStackProps) {
    super(scope, id, props);

    this.lambdaCode = Code.fromCfnParameters();

    const lambdaRole = new Role(this, "lambda-role", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      roleName: "ApplicationStackLambdaRole",
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName("AWSPriceListServiceFullAccess"),
      ],
    });

    const handler = new Function(
      this,
      `${props.applicationName}-${props.stage}`,
      {
        runtime: Runtime.NODEJS_14_X,
        code: this.lambdaCode,
        handler: "dist/lambda.handler",
        functionName: `${props.applicationName}-${props.stage}`,
        role: lambdaRole,
      }
    );

    const api = new LambdaRestApi(
      this,
      `${props.applicationName}-api-${props.stage}`,
      {
        handler,
        proxy: false,
        restApiName: `${props.applicationName}-api-${props.stage}`,
      }
    );

    // root
    api.root.addMethod("GET");

    // graphql
    const graphql = api.root.addResource("graphql");
    graphql.addMethod("ANY");

    // api
    const apiRoot = api.root.addResource("api");

    // lambda
    const lambda = apiRoot.addResource("lambda");
    const lambdaUsageType = lambda.addResource("{usageType}");
    lambdaUsageType.addMethod("GET");

    // ec2
    const ec2 = apiRoot.addResource("ec2");
    const ec2InstanceType = ec2.addResource("{instanceType}");
    const ec2region = ec2InstanceType.addResource("{region}");
    ec2region.addMethod("GET");

    new CfnOutput(this, "apiPath", {
      value: api.root.path,
      description: "Path of the API",
    });

    const dashboard = new Dashboard(
      this,
      `${props.applicationName}-dashboard`,
      { dashboardName: `${props.applicationName}-dashboard` }
    );
    dashboard.addWidgets(
      // Lambda
      this.buildWidget("Duration", "AWS/Lambda"),
      this.buildWidget("Throttle", "AWS/Lambda"),
      this.buildWidget("Invocations", "AWS/Lambda"),
      this.buildWidget("ConcurrentExecutions", "AWS/Lambda"),
      this.buildWidget("Errors", "AWS/Lambda"),

      // API Gateway
      this.buildWidget("5XXError", "AWS/ApiGateway"),
      this.buildWidget("4XXError", "AWS/ApiGateway"),
      this.buildWidget("IntegrationLatency", "AWS/ApiGateway"),
      this.buildWidget("Latency", "AWS/ApiGateway"),
      this.buildWidget("Count", "AWS/ApiGateway")
    );
  }

  buildWidget(metricName: string, namespace: string): GraphWidget {
    return new GraphWidget({
      width: 24,
      title: metricName,
      left: [
        new Metric({
          namespace,
          metricName,
          label: metricName,
          statistic: "avg",
        }),
      ],
    });
  }
}
