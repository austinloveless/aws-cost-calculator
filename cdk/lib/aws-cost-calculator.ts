import { Stack, Construct, CfnOutput } from "@aws-cdk/core";
import { LambdaRestApi } from "@aws-cdk/aws-apigateway";
import { Function, Runtime, Code, IFunction } from "@aws-cdk/aws-lambda";

export class AWSCostCalculator extends Stack {
  readonly handler: IFunction;

  constructor(scope: Construct, id: string, props?: undefined) {
    super(scope, id, props);

    const handler = new Function(this, "aws-cost-calculator", {
      runtime: Runtime.NODEJS_14_X,
      code: Code.fromAsset("src"),
      handler: "dist/lambda.handler",
      functionName: "aws-cost-calculator",
    });

    const api = new LambdaRestApi(this, "aws-cost-calculator-api", {
      handler,
      proxy: false,
      restApiName: "aws-cost-calculator-api",
    });

    api.root.addMethod("GET");
    const github = api.root.addResource("github");
    github.addMethod("GET");

    new CfnOutput(this, "apiPath", {
      value: api.root.path,
      description: "Path of the API",
    });
  }
}
