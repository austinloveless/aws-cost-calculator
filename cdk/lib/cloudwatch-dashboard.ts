import { Dashboard } from "@aws-cdk/aws-cloudwatch";
import { GraphWidget, Metric } from "@aws-cdk/aws-cloudwatch";
import { Construct, NestedStack, StackProps } from "@aws-cdk/core";

interface CloudwatchStackProps extends StackProps {
  applicationName: string;
  stage: string;
}

export class CloudWatchDashboard extends NestedStack {
  constructor(scope: Construct, id: string, props: CloudwatchStackProps) {
    super(scope, id);

    const dashboard = new Dashboard(
      this,
      `${props.applicationName}-dashboard`,
      { dashboardName: `${props.applicationName}-dashboard` }
    );
    dashboard.addWidgets(
      // Lambda
      this.buildWidget("Duration", "AWS/Lambda"),
      this.buildWidget("Throttle", "AWS/Lambda"),
      this.buildWidget("Invocations", "AWS/Lambda"),
      this.buildWidget("ConcurrentExecutions", "AWS/Lambda"),
      this.buildWidget("Errors", "AWS/Lambda"),

      // API Gateway
      this.buildWidget("5XXError", "AWS/ApiGateway"),
      this.buildWidget("4XXError", "AWS/ApiGateway"),
      this.buildWidget("IntegrationLatency", "AWS/ApiGateway"),
      this.buildWidget("Latency", "AWS/ApiGateway"),
      this.buildWidget("Count", "AWS/ApiGateway")
    );
  }

  buildWidget(metricName: string, namespace: string): GraphWidget {
    return new GraphWidget({
      width: 24,
      title: metricName,
      left: [
        new Metric({
          namespace,
          metricName,
          label: metricName,
          statistic: "avg",
        }),
      ],
    });
  }
}
