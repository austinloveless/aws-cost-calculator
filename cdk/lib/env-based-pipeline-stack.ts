import { Role, AccountPrincipal, PolicyStatement } from "@aws-cdk/aws-iam";
import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  CfnOutput,
  CfnCapabilities,
} from "@aws-cdk/core";

import { ApplicationStack } from "./application-stack";
import {
  BuildSpec,
  LinuxBuildImage,
  PipelineProject,
} from "@aws-cdk/aws-codebuild";
import {
  CloudFormationCreateUpdateStackAction,
  CodeBuildAction,
  GitHubSourceAction,
} from "@aws-cdk/aws-codepipeline-actions";
import { Artifact, Pipeline } from "@aws-cdk/aws-codepipeline";
import { Secret } from "@aws-cdk/aws-secretsmanager";
import { Key } from "@aws-cdk/aws-kms";
import { Bucket, BucketEncryption } from "@aws-cdk/aws-s3";

export interface PipelineStackProps extends StackProps {
  readonly devApplicationStack: ApplicationStack;
  readonly stageApplicationStack: ApplicationStack;
  readonly prodApplicationStack: ApplicationStack;
  readonly rootAccountId: string;
  readonly devAccountId: string;
  readonly stageAccountId: string;
  readonly prodAccountId: string;
  readonly gitHubOwner: string;
  readonly gitHubRepo: string;
}

export class EnvironmentBasedPipelineStack extends Stack {
  constructor(app: App, id: string, props: PipelineStackProps) {
    super(app, id, props);

    const gitHubSecret = Secret.fromSecretCompleteArn(
      this,
      "github-secret",
      `arn:aws:secretsmanager:us-east-1:${props.rootAccountId}:secret:github-secret-bEZ24D`
    );

    const sourceOutput = new Artifact();

    //dev
    const devSourceAction = new GitHubSourceAction({
      actionName: "GitHub_Source",
      owner: props.gitHubOwner,
      repo: props.gitHubRepo,
      branch: "develop",
      oauthToken: gitHubSecret.secretValue,
      output: sourceOutput,
    });
    const devDeploymentRole = Role.fromRoleArn(
      this,
      "DevDeploymentRole",
      `arn:aws:iam::${props.devAccountId}:role/CloudFormationDeploymentRole`,
      {
        mutable: false,
      }
    );
    const devCrossAccountRole = Role.fromRoleArn(
      this,
      "DevCrossAccountRole",
      `arn:aws:iam::${props.devAccountId}:role/CodePipelineCrossAccountRole`,
      {
        mutable: false,
      }
    );

    //stage
    const stageSourceAction = new GitHubSourceAction({
      actionName: "GitHub_Source",
      owner: props.gitHubOwner,
      repo: props.gitHubRepo,
      branch: "stage",
      oauthToken: gitHubSecret.secretValue,
      output: sourceOutput,
    });
    const stageDeploymentRole = Role.fromRoleArn(
      this,
      "StageDeploymentRole",
      `arn:aws:iam::${props.stageAccountId}:role/CloudFormationDeploymentRole`,
      {
        mutable: false,
      }
    );
    const stageCrossAccountRole = Role.fromRoleArn(
      this,
      "StageCrossAccountRole",
      `arn:aws:iam::${props.stageAccountId}:role/CodePipelineCrossAccountRole`,
      {
        mutable: false,
      }
    );

    //prod
    const prodSourceAction = new GitHubSourceAction({
      actionName: "GitHub_Source",
      owner: props.gitHubOwner,
      repo: props.gitHubRepo,
      branch: "main",
      oauthToken: gitHubSecret.secretValue,
      output: sourceOutput,
    });
    const prodDeploymentRole = Role.fromRoleArn(
      this,
      "ProdDeploymentRole",
      `arn:aws:iam::${props.prodAccountId}:role/CloudFormationDeploymentRole`,
      {
        mutable: false,
      }
    );

    const prodCrossAccountRole = Role.fromRoleArn(
      this,
      "ProdCrossAccountRole",
      `arn:aws:iam::${props.prodAccountId}:role/CodePipelineCrossAccountRole`,
      {
        mutable: false,
      }
    );

    // dev
    const devAccountRootPrincipal = new AccountPrincipal(props.devAccountId);

    //stage
    const stageAccountRootPrincipal = new AccountPrincipal(
      props.stageAccountId
    );

    //prod
    const prodAccountRootPrincipal = new AccountPrincipal(props.prodAccountId);

    const key = new Key(this, "ArtifactKey", {
      alias: "key/artifact-key",
    });

    //dev
    key.grantDecrypt(devAccountRootPrincipal);
    key.grantDecrypt(devCrossAccountRole);

    //stage
    key.grantDecrypt(stageAccountRootPrincipal);
    key.grantDecrypt(stageCrossAccountRole);

    //prod
    key.grantDecrypt(prodAccountRootPrincipal);
    key.grantDecrypt(prodCrossAccountRole);

    const artifactBucket = new Bucket(this, "ArtifactBucket", {
      bucketName: `env-based-artifact-bucket-${this.account}`,
      removalPolicy: RemovalPolicy.DESTROY,
      encryption: BucketEncryption.KMS,
      encryptionKey: key,
    });

    // dev
    artifactBucket.grantPut(devAccountRootPrincipal);
    artifactBucket.grantRead(devAccountRootPrincipal);

    // stage
    artifactBucket.grantPut(stageAccountRootPrincipal);
    artifactBucket.grantRead(stageAccountRootPrincipal);

    // prod
    artifactBucket.grantPut(prodAccountRootPrincipal);
    artifactBucket.grantRead(prodAccountRootPrincipal);

    const cdkBuild = new PipelineProject(this, "CdkBuild", {
      buildSpec: BuildSpec.fromObject({
        version: "0.2",
        phases: {
          install: {
            commands: ["npm install", "cd src", "npm install", "cd ../"],
          },
          build: {
            commands: ["npm run build", "npx cdk synth --all -o dist"],
          },
        },
        artifacts: {
          "base-directory": "dist",
          files: ["*ApplicationStack*"],
        },
      }),
      environment: {
        buildImage: LinuxBuildImage.AMAZON_LINUX_2_2,
      },
      encryptionKey: key,
    });
    const lambdaBuild = new PipelineProject(this, "ApplicationBuild", {
      buildSpec: BuildSpec.fromObject({
        version: "0.2",
        phases: {
          install: {
            commands: [
              "npm install",
              "cd src",
              "npm install",
              "npm install -g typescript",
            ],
          },
          pre_build: {
            commands: ["cd ../", "npm run test", "cd src"],
          },
          build: {
            commands: ["npm run build"],
          },
        },
        artifacts: {
          "base-directory": "src",
          files: ["dist/**/*", "node_modules/**/*"],
        },
      }),
      environment: {
        buildImage: LinuxBuildImage.AMAZON_LINUX_2_2,
      },
      encryptionKey: key,
    });

    const cdkBuildOutput = new Artifact("CdkBuildOutput");
    const lambdaBuildOutput = new Artifact("LambdaBuildOutput");

    // Develop Branch Pipeline
    const devPipeline = new Pipeline(this, "DevPipeline", {
      pipelineName: "DevCrossAccountPipeline",
      artifactBucket: artifactBucket,
      stages: [
        {
          stageName: "Source",
          actions: [devSourceAction],
        },
        {
          stageName: "Build",
          actions: [
            new CodeBuildAction({
              actionName: "Application_Build",
              project: lambdaBuild,
              input: sourceOutput,
              outputs: [lambdaBuildOutput],
            }),
            new CodeBuildAction({
              actionName: "CDK_Synth",
              project: cdkBuild,
              input: sourceOutput,
              outputs: [cdkBuildOutput],
            }),
          ],
        },
        {
          stageName: "dev",
          actions: [
            new CloudFormationCreateUpdateStackAction({
              actionName: "Deploy",
              templatePath: cdkBuildOutput.atPath(
                "DevApplicationStack.template.json"
              ),
              stackName: "DevApplicationDeploymentStack",
              adminPermissions: true,
              parameterOverrides: {
                ...props.devApplicationStack.lambdaCode.assign(
                  lambdaBuildOutput.s3Location
                ),
              },
              deploymentRole: devDeploymentRole,
              cfnCapabilities: [
                CfnCapabilities.ANONYMOUS_IAM,
                CfnCapabilities.NAMED_IAM,
              ],
              extraInputs: [lambdaBuildOutput],
              role: devCrossAccountRole,
            }),
          ],
        },
      ],
    });

    devPipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["pricing:*"],
        resources: ["*"],
      })
    );
    devPipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["sts:AssumeRole"],
        resources: [`arn:aws:iam::${props.devAccountId}:role/*`],
      })
    );

    // Stage Branch Pipeline
    const stagePipeline = new Pipeline(this, "StagePipeline", {
      pipelineName: "StageCrossAccountPipeline",
      artifactBucket: artifactBucket,
      stages: [
        {
          stageName: "Source",
          actions: [stageSourceAction],
        },
        {
          stageName: "Build",
          actions: [
            new CodeBuildAction({
              actionName: "Application_Build",
              project: lambdaBuild,
              input: sourceOutput,
              outputs: [lambdaBuildOutput],
            }),
            new CodeBuildAction({
              actionName: "CDK_Synth",
              project: cdkBuild,
              input: sourceOutput,
              outputs: [cdkBuildOutput],
            }),
          ],
        },
        {
          stageName: "stage",
          actions: [
            new CloudFormationCreateUpdateStackAction({
              actionName: "Deploy",
              templatePath: cdkBuildOutput.atPath(
                "StageApplicationStack.template.json"
              ),
              stackName: "StageApplicationDeploymentStack",
              adminPermissions: true,
              parameterOverrides: {
                ...props.stageApplicationStack.lambdaCode.assign(
                  lambdaBuildOutput.s3Location
                ),
              },
              deploymentRole: stageDeploymentRole,
              cfnCapabilities: [
                CfnCapabilities.ANONYMOUS_IAM,
                CfnCapabilities.NAMED_IAM,
              ],
              extraInputs: [lambdaBuildOutput],
              role: stageCrossAccountRole,
            }),
          ],
        },
      ],
    });

    stagePipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["pricing:*"],
        resources: ["*"],
      })
    );
    stagePipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["sts:AssumeRole"],
        resources: [`arn:aws:iam::${props.stageAccountId}:role/*`],
      })
    );

    // Prod Branch Pipeline
    const prodPipeline = new Pipeline(this, "ProdPipeline", {
      pipelineName: "ProdCrossAccountPipeline",
      artifactBucket: artifactBucket,
      stages: [
        {
          stageName: "Source",
          actions: [prodSourceAction],
        },
        {
          stageName: "Build",
          actions: [
            new CodeBuildAction({
              actionName: "Application_Build",
              project: lambdaBuild,
              input: sourceOutput,
              outputs: [lambdaBuildOutput],
            }),
            new CodeBuildAction({
              actionName: "CDK_Synth",
              project: cdkBuild,
              input: sourceOutput,
              outputs: [cdkBuildOutput],
            }),
          ],
        },
        {
          stageName: "prod",
          actions: [
            new CloudFormationCreateUpdateStackAction({
              actionName: "Deploy",
              templatePath: cdkBuildOutput.atPath(
                "ProdApplicationStack.template.json"
              ),
              stackName: "ProdApplicationDeploymentStack",
              adminPermissions: true,
              parameterOverrides: {
                ...props.prodApplicationStack.lambdaCode.assign(
                  lambdaBuildOutput.s3Location
                ),
              },
              deploymentRole: prodDeploymentRole,
              cfnCapabilities: [
                CfnCapabilities.ANONYMOUS_IAM,
                CfnCapabilities.NAMED_IAM,
              ],
              extraInputs: [lambdaBuildOutput],
              role: prodCrossAccountRole,
            }),
          ],
        },
      ],
    });

    prodPipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["pricing:*"],
        resources: ["*"],
      })
    );
    prodPipeline.addToRolePolicy(
      new PolicyStatement({
        actions: ["sts:AssumeRole"],
        resources: [`arn:aws:iam::${props.prodAccountId}:role/*`],
      })
    );

    new CfnOutput(this, "ArtifactBucketEncryptionKeyArn", {
      value: key.keyArn,
      exportName: "ArtifactBucketEncryptionKey",
    });
  }
}
