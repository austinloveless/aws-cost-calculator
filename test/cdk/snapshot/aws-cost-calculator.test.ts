import { SynthUtils } from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";

import { AWSCostCalculator } from "../../../cdk/lib/aws-cost-calculator";

test("aws-cost-calculator stack", () => {
  const app = new cdk.App();
  const stack = new AWSCostCalculator(app, "aws-cost-calculator");
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
