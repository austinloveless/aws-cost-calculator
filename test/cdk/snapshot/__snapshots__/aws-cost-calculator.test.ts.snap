// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws-cost-calculator stack 1`] = `
Object {
  "Outputs": Object {
    "apiPath": Object {
      "Description": "Path of the API",
      "Value": "/",
    },
    "awscostcalculatorapiEndpointCEB131A5": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "awscostcalculatorapiCA13CAA1",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "awscostcalculatorapiDeploymentStageprod2074B5E2",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86ArtifactHash42B6A961": Object {
      "Description": "Artifact hash for asset \\"db433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86\\"",
      "Type": "String",
    },
    "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86S3BucketCB52BBBC": Object {
      "Description": "S3 bucket for asset \\"db433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86\\"",
      "Type": "String",
    },
    "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86S3VersionKeyF0D27919": Object {
      "Description": "S3 key for asset version \\"db433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "awscostcalculator1974DCEF": Object {
      "DependsOn": Array [
        "awscostcalculatorServiceRole48262EC8",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86S3BucketCB52BBBC",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86S3VersionKeyF0D27919",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersdb433b35df5c6b35a6c06fd7ea0b9dd9b5cada1c9010985385f2548ebbeb6a86S3VersionKeyF0D27919",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "aws-cost-calculator",
        "Handler": "dist/lambda.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "awscostcalculatorServiceRole48262EC8",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "awscostcalculatorServiceRole48262EC8": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "awscostcalculatorapiAccount0214CA2F": Object {
      "DependsOn": Array [
        "awscostcalculatorapiCA13CAA1",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "awscostcalculatorapiCloudWatchRoleF7FE395A",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "awscostcalculatorapiCA13CAA1": Object {
      "Properties": Object {
        "Name": "aws-cost-calculator-api",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "awscostcalculatorapiCloudWatchRoleF7FE395A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "awscostcalculatorapiDeploymentF6E916ADd9c62d99fc8311e399dd99751abb96d4": Object {
      "DependsOn": Array [
        "awscostcalculatorapiGETE769F160",
        "awscostcalculatorapigraphqlANYA9F463B3",
        "awscostcalculatorapigraphqlBBB1800C",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "awscostcalculatorapiCA13CAA1",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "awscostcalculatorapiDeploymentStageprod2074B5E2": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "awscostcalculatorapiDeploymentF6E916ADd9c62d99fc8311e399dd99751abb96d4",
        },
        "RestApiId": Object {
          "Ref": "awscostcalculatorapiCA13CAA1",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "awscostcalculatorapiGETApiPermissionTestawscostcalculatorawscostcalculatorapiDD133946GETBE46120A": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "awscostcalculator1974DCEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "awscostcalculatorapiCA13CAA1",
              },
              "/test-invoke-stage/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "awscostcalculatorapiGETApiPermissionawscostcalculatorawscostcalculatorapiDD133946GETA3AA3713": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "awscostcalculator1974DCEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "awscostcalculatorapiCA13CAA1",
              },
              "/",
              Object {
                "Ref": "awscostcalculatorapiDeploymentStageprod2074B5E2",
              },
              "/GET/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "awscostcalculatorapiGETE769F160": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "GET",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "awscostcalculator1974DCEF",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "awscostcalculatorapiCA13CAA1",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "awscostcalculatorapiCA13CAA1",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "awscostcalculatorapigraphqlANYA9F463B3": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "awscostcalculator1974DCEF",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": Object {
          "Ref": "awscostcalculatorapigraphqlBBB1800C",
        },
        "RestApiId": Object {
          "Ref": "awscostcalculatorapiCA13CAA1",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "awscostcalculatorapigraphqlANYApiPermissionTestawscostcalculatorawscostcalculatorapiDD133946ANYgraphql03458E97": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "awscostcalculator1974DCEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "awscostcalculatorapiCA13CAA1",
              },
              "/test-invoke-stage/*/graphql",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "awscostcalculatorapigraphqlANYApiPermissionawscostcalculatorawscostcalculatorapiDD133946ANYgraphql74EDB646": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "awscostcalculator1974DCEF",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "awscostcalculatorapiCA13CAA1",
              },
              "/",
              Object {
                "Ref": "awscostcalculatorapiDeploymentStageprod2074B5E2",
              },
              "/*/graphql",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "awscostcalculatorapigraphqlBBB1800C": Object {
      "Properties": Object {
        "ParentId": Object {
          "Fn::GetAtt": Array [
            "awscostcalculatorapiCA13CAA1",
            "RootResourceId",
          ],
        },
        "PathPart": "graphql",
        "RestApiId": Object {
          "Ref": "awscostcalculatorapiCA13CAA1",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
  },
}
`;
